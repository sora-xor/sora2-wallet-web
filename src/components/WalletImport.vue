<template>
  <wallet-base :title="t('import.title')" show-back @back="handleBack">
    <div class="wallet-import">
      <template v-if="step === 1">
        <s-select :placeholder="t('import.sourceType.placeholder')" v-model="sourceType" border-radius="mini">
          <s-option
            v-for="type in SourceTypes"
            :key="type"
            :value="type"
            :label="t(`import.sourceType.${type}`)"
          />
        </s-select>
        <template v-if="sourceType !== SourceTypes.PolkadotJs">
          <s-input type="textarea" :placeholder="t(`import.${sourceType}.placeholder`)" v-model="seed" border-radius="mini" />
          <div class="wallet-import-hint">{{ t(`import.${sourceType}.hint`) }}</div>
        </template>
      </template>
      <template v-else>
        <template v-if="sourceType !== SourceTypes.PolkadotJs">
          <wallet-account :name="importFormData.name" />
          <s-input
            :placeholder="t(`import.${SourceTypes.MnemonicSeed}.name.placeholder`)"
            v-model="importFormData.name"
            border-radius="mini"
          />
          <div class="wallet-import-hint">{{ t(`import.${SourceTypes.MnemonicSeed}.name.hint`) }}</div>
          <s-input
            show-password
            :placeholder="t(`import.${SourceTypes.MnemonicSeed}.password.placeholder`)"
            v-model="importFormData.password"
            border-radius="mini"
          />
          <div>
            <div class="wallet-import-condition" v-for="condition in PasswordConditions" :key="condition.title">
              <s-icon name="check-mark" :class="{ 'error': !condition.regexp.test(importFormData.password) }" />
              <span>{{ t(`import.${SourceTypes.MnemonicSeed}.password.${condition.title}`) }}</span>
            </div>
          </div>
          <s-input
            show-password
            :placeholder="t(`import.${SourceTypes.MnemonicSeed}.repeatedPassword.placeholder`)"
            v-model="importFormData.repeatedPassword"
            border-radius="mini"
          />
          <div class="wallet-import-hint">{{ t(`import.${SourceTypes.MnemonicSeed}.password.hint`) }}</div>
        </template>
        <s-select
          v-else
          :placeholder="t(`import.${SourceTypes.PolkadotJs}.selectAccount`)"
          v-model="selectedPolkadotJsAccount"
          border-radius="mini"
        >
          <s-option
            v-for="account in polkadotJsAccounts"
            :key="account.address"
            :value="account.address"
            :label="account.name"
          />
        </s-select>
      </template>
      <s-button
        type="primary"
        size="medium"
        :loading="loading"
        :disabled="disabled"
        @click="handleImport"
      >
        {{ t('import.action') }}
      </s-button>
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Action, Getter } from 'vuex-class'
import last from 'lodash/fp/last'

import TranslationMixin from './mixins/TranslationMixin'
import LoadingMixin from './mixins/LoadingMixin'
import WalletBase from './WalletBase.vue'
import WalletAccount from './WalletAccount.vue'
import { RouteNames, SourceTypes, PasswordConditions } from '../consts'
import { checkValidSeed } from '../api/account'

@Component({
  components: { WalletBase, WalletAccount }
})
export default class WalletImport extends Mixins(TranslationMixin, LoadingMixin) {
  readonly SourceTypes = SourceTypes
  readonly PasswordConditions = PasswordConditions

  @Action navigate
  @Action login
  @Action getAddress
  @Action importPolkadotJs
  @Action getPolkadotJsAccounts

  sourceType = SourceTypes.PolkadotJs
  polkadotJsAccounts: Array<{ address: string; name: string }> = []
  selectedPolkadotJsAccount: any = ''
  seed = ''
  step = 1
  importFormData = {
    name: '',
    password: '',
    repeatedPassword: ''
  }

  get disabled (): boolean {
    if (this.sourceType === SourceTypes.PolkadotJs) {
      return false
    }
    return this.step === 1
      ? !(this.seed && checkValidSeed(this.seed))
      : (
        Object.values(this.importFormData).some(prop => !prop) ||
        PasswordConditions.map(({ regexp }) => regexp).some(regexp => !regexp.test(this.importFormData.password)) ||
        this.importFormData.password !== this.importFormData.repeatedPassword
      )
  }

  handleBack (): void {
    if (this.step === 1) {
      this.navigate({ name: RouteNames.WalletConnection })
      return
    }
    this.step = 1
  }

  async handleImport (): Promise<void> {
    const isInternalImport = [SourceTypes.MnemonicSeed, SourceTypes.RawSeed].includes(this.sourceType)
    try {
      await this.withApi(async () => {
        if (this.step === 1) {
          if (isInternalImport) {
            this.getAddress({ seed: this.seed })
          } else {
            this.polkadotJsAccounts = await this.getPolkadotJsAccounts()
            this.selectedPolkadotJsAccount = (last(this.polkadotJsAccounts) || {}).address
          }
        } else {
          if (isInternalImport) {
            const { name, password } = this.importFormData
            await this.login({ name, password, seed: this.seed })
          } else {
            await this.importPolkadotJs({ address: this.selectedPolkadotJsAccount })
          }
        }
      })
    } catch (error) {
      this.$alert(this.t((error as Error).message), this.t('errorText'))
      if (this.sourceType === SourceTypes.PolkadotJs && this.step === 2) {
        this.step = 1
      }
      return
    }
    if (this.step === 1) {
      this.step = 2
      return
    }
    this.navigate({ name: RouteNames.Wallet })
  }
}
</script>

<style lang="scss">
@include select-icon('wallet-import');
</style>

<style scoped lang="scss">
$valid-icon-size: 6px;

.wallet-import {
  &-hint,
  &-condition {
    @include hint-text(true);
  }
  &-condition {
    display: inline-flex;
    align-items: center;
    width: 100%;
    .s-icon-check-mark {
      display: block;
      margin-left: $valid-icon-size;
      margin-right: $valid-icon-size;
      width: $valid-icon-size * 2;
      font-weight: bold;
      color: var(--s-color-status-success);
      &.error {
        position: relative;
        &:before {
          content: "";
          position: absolute;
          display: block;
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
          margin: auto;
          background-color: var(--s-color-base-content-tertiary);
          border-radius: 50%;
          height: $valid-icon-size;
          width: $valid-icon-size;
        }
      }
    }
  }
  > button {
    width: 100%;
  }
  > :not(:last-child) {
    margin-bottom: $basic-spacing;
  }
}
</style>
