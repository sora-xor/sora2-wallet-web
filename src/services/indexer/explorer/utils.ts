import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
