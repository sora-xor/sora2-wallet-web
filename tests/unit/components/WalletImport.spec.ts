import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import WalletImport from '@/components/WalletImport.vue'
import { TranslationMock, SoramitsuElementsImport } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Vuex.Store({
  modules: {
    Account: {
      actions: {
        login: jest.fn(),
        importPolkadotJs: jest.fn(),
        getPolkadotJsAccounts: jest.fn(),
        getAddress: jest.fn()
      }
    },
    Router: {
      actions: {
        navigate: jest.fn()
      }
    }
  } as any
})

describe('WalletImport.vue', () => {
  beforeEach(() => {
    SoramitsuElementsImport(localVue)
    TranslationMock(WalletImport)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(WalletImport, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
