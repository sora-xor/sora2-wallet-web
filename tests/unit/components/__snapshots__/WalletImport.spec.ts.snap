// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletImport.vue should renders correctly 1`] = `
<wallet-base-stub
  showback="true"
  title=""
>
  <div
    class="wallet-import"
  >
    <s-select-stub
      autocomplete="off"
      borderradius="mini"
      inputtype="input"
      loadingtext="Loading"
      multiplelimit="0"
      multipletextprefix=""
      name=""
      nodatatext="No data"
      placeholder="Select"
      popperclass=""
      value="MnemonicSeed"
    >
      <s-option-stub
        label=""
        value="MnemonicSeed"
      />
      <s-option-stub
        label=""
        value="RawSeed"
      />
      <s-option-stub
        label=""
        value="PolkadotJs"
      />
    </s-select-stub>
     
    <s-input-stub
      accept=""
      autocomplete="off"
      borderradius="mini"
      form=""
      label=""
      name=""
      placeholder=""
      prefix=""
      size="big"
      suffix=""
      tabindex=""
      type="textarea"
      value=""
    />
     
    <div
      class="wallet-import-hint"
    >
      
    </div>
     
    <s-button-stub
      borderradius="small"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
    >
      
      
    
    </s-button-stub>
  </div>
</wallet-base-stub>
`;
